crates/hir-ty/src/method_resolution.rs

```diff
@@ -575,6 +576,59 @@ pub(crate) fn iterate_method_candidates<T>(
     slot
 }
 
+pub fn lookup_impl_method(
+    self_ty: &Ty,
+    db: &dyn HirDatabase,
+    env: Arc<TraitEnvironment>,
+    trait_: TraitId,
+    name: &Name,
+) -> Option<FunctionId> {
+    let self_ty_fp = TyFingerprint::for_trait_impl(self_ty)?;
+    let trait_impls = TraitImpls::trait_impls_in_deps_query(db, env.krate);
+    let impls = trait_impls.for_trait_and_self_ty(trait_, self_ty_fp);
+    let mut table = InferenceTable::new(db, env.clone());
+    find_matching_impl(impls, &mut table, &self_ty).and_then(|data| {
+        data.items.iter().find_map(|it| match it {
+            AssocItemId::FunctionId(f) => (db.function_data(*f).name == *name).then(|| *f),
+            _ => None,
+        })
+    })
+}
+
+fn find_matching_impl(
+    mut impls: impl Iterator<Item = ImplId>,
+    table: &mut InferenceTable,
+    self_ty: &Ty,
+) -> Option<Arc<ImplData>> {
+    let db = table.db;
+    loop {
+        let impl_ = impls.next()?;
+        let r = table.run_in_snapshot(|table| {
+            let impl_data = db.impl_data(impl_);
+            let substs =
+                TyBuilder::subst_for_def(db, impl_).fill_with_inference_vars(table).build();
+            let impl_ty = db.impl_self_ty(impl_).substitute(Interner, &substs);
+
+            table
+                .unify(self_ty, &impl_ty)
+                .then(|| {
+                    let wh_goals =
+                        crate::chalk_db::convert_where_clauses(db, impl_.into(), &substs)
+                            .into_iter()
+                            .map(|b| b.into_well_formed_goal(Interner).cast(Interner));
```

Comment:
actually looking at this again, I don't think `into_well_formed_goal` is the right thing here -- the where clauses should be the goals themselves directly

Reply:
you means goal::Holds? i i have a confused. it’s looks like well-formed is a subset fromenv 🤔

Reply:
I think we just want `b.cast(Interner)`. `QuantifiedWhereClause` can be directly cast into a goal, the `into_well_formed_goal` and `into_from_env_goal` are just for special cases.

Reply:
as i see, wc can convert to 3 DomainGoal: Holds, wf and fromenv. if b.cast to Holds, as i try it can not work well. And as chalk docs, it uses wf to solve implied bound, if i not misunderstood.

Reply:
I'm not sure what you're trying to say. Chalk uses `CastTo<Goal>` (i.e. `Holds`, yes) to check impls. This is unrelated to implied bounds or well-formedness.

Reply:
(Well-formedness of the substituted where clauses would *require* that they hold, but well-formedness is more than we want/need to check here.)

Reply:
```
trait G {
    fn g(&self);
}
trait BParent{}
trait Bound: BParent{}
struct Gen<T>(T);
impl <T> G for Gen<T>
where T : Bound
{
    fn g(&self){
     //^
    }
}
struct A;
impl Bound for A{}
fn f() {
    let gen = Gen::<A>(A);
    gen.g$0();
}
```

like this, I no impl `BParent` for `A`. if use `b.cast`, passed.

Reply:
It'd be good to add that to the test cases. IMO it's correct that this passes. That's the point of the differentiation between WellFormed and Holds: We check for each impl that it is well-formed, and then we can assume in all other places that it is the case and don't need to re-check it again and again. The `impl Bound for A {}` in this case is not well-formed, but we don't want to check that for every unrelated impl.

Reply:
👌